File: 1472-design-browser-history/1472-design-browser-history.js
/**
 * @param {string} homepage
 */
var BrowserHistory = function(homepage) {
  this.history = [homepage];
  this.future = [];
};

/** 
 * @param {string} url
 * @return {void}
 */
BrowserHistory.prototype.visit = function(url) {
  this.history.push(url);
  this.future = [];
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.back = function(steps) {
  while (steps-- && this.history.length > 1) 
    this.future.push(this.history.pop());
  
  return this.history.at(-1);
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.forward = function(steps) {
  while (steps-- && this.future.length)
    this.history.push(this.future.pop());
  
  return this.history.at(-1);
};

/** 
 * Your BrowserHistory object will be instantiated and called as such:
 * var obj = new BrowserHistory(homepage)
 * obj.visit(url)
 * var param_2 = obj.back(steps)
 * var param_3 = obj.forward(steps)
 */

File: 1472-design-browser-history/1472-design-browser-history.py
class BrowserHistory:

    def __init__(self, homepage: str):
        self.history = [homepage]
        self.future = []

    def visit(self, url: str) -> None:
        self.history.append(url)
        self.future = []

    def back(self, steps: int) -> str:
        while steps and len(self.history) > 1:
            top = self.history.pop()
            self.future.append(top)
            steps -= 1
        return self.history[-1]

    def forward(self, steps: int) -> str:
        while steps and len(self.future):
            top = self.future.pop()
            self.history.append(top)
            steps -= 1
        return self.history[-1]


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)